src/krb5_plugin/common/radius_kdcpreauth.c

```diff
@@ -0,0 +1,609 @@
+    state.length = cookie.length - sizeof(uint16_t);
+
+    *_state = state;
+
+    return 0;
+}
+
+
+/* Some attributes have limited length. In order to accept longer values,
+ * we will concatenate all attribute values to single krb5_data. */
+krb5_error_code
+sss_radiuskdc_get_complete_attr(const krad_packet *rres,
+                                const char *attr_name,
+                                krb5_data *_data)
+{
+    krad_attr attr = krad_attr_name2num(attr_name);
+    const krb5_data *rmsg;
+    krb5_data data = {0};
+    unsigned int memindex;
+    unsigned int i;
+
+    i = 0;
+    do {
+        rmsg = krad_packet_get_attr(rres, attr, i);
+        if (rmsg != NULL) {
+            data.length += rmsg->length;
+        }
+        i++;
+    } while (rmsg != NULL);
+
+    if (data.length == 0) {
+        return ENOENT;
+    }
+
+    data.data = malloc(data.length);
+    if (data.data == NULL) {
+        return ENOMEM;
+    }
+
+    i = 0;
+    memindex = 0;
+    do {
+        rmsg = krad_packet_get_attr(rres, attr, i);
+        if (rmsg != NULL) {
+            memcpy(&data.data[memindex], rmsg->data, rmsg->length);
+            memindex += rmsg->length;
+        }
+        i++;
+    } while (rmsg != NULL);
+
+    if (memindex != data.length) {
+        free(data.data);
+        return ERANGE;
+    }
+
+    *_data = data;
+
+    return 0;
+}
+
+/* From krad internals, RFC 2865 */
+#ifndef UCHAR_MAX
+#define UCHAR_MAX 255
+#endif
+#define MAX_ATTRSIZE (UCHAR_MAX - 2)
+
+krb5_error_code
+sss_radiuskdc_put_complete_attr(krad_attrset *attrset,
+                                krad_attr attr,
+                                const krb5_data *datap)
+{
+    krb5_data state = {0};
+    char *p = datap->data;
+    unsigned int len = datap->length;
+    krb5_error_code ret = 0;
+
+    do {
+        /* - 5 to make sure we fit into minimal value length */
+        state.data = p;
+        state.length = MIN(MAX_ATTRSIZE - 5, len);
+        p += state.length;
+
+        ret = krad_attrset_add(attrset, attr, &(state));
+        if (ret != 0) {
+            break;
+        }
+        len -= state.length;
+    } while (len > 0);
+
+    return ret;
+}
+
+char *
+sss_radiuskdc_get_attr_as_string(const krad_packet *packet, const char *attr)
+{
+    krb5_data data = {0};
+    krb5_error_code ret;
+    char *str;
+
+    ret = sss_radiuskdc_get_complete_attr(packet, attr, &data);
+    if (ret != 0) {
+        return NULL;
+    }
+
+    str = strndup(data.data, data.length);
+    free(data.data);
+
+    return str;
+}
+
+krb5_error_code
+sss_radiuskdc_set_attr_as_string(krad_attrset *attrset,
+                                 const char *attr,
+                                 const char *value)
+{
+    krb5_data data = {0};
+    krb5_error_code ret;
+
+    data.data = discard_const(value);
+    data.length = strlen(value) + 1;
+
+    ret = sss_radiuskdc_put_complete_attr(attrset,
+                                          krad_code_name2num(attr),
```

Comment:
`krad_attr_name2num()` is needed here to translate attribute names like `Proxy-State`.